File: Tree/BST/BST.java
class BST {
    static class Node {
        int data;
        Node left;
        Node right;

        Node(int data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }

    public static Node insert(Node root, int val) {
        if (root == null) {
            root = new Node(val);
            return root;
        }
        if (val < root.data) {
            root.left = insert(root.left, val);
        } else {
            root.right = insert(root.right, val);
        }
        return root;
    }

    public static void inOrder(Node root) {
        if (root == null) return;
        inOrder(root.left);
        System.out.print(root.data + " ");
        inOrder(root.right);
    }
    public static boolean search(Node root, int key){
        if(root == null)return false;
        if(root.data> key){
            return search(root.left,key);
        }else if(root.data==key){
            return true;
        }
        else{
            return search(root.right,key);
        }
    }
    public static Node deleteEle(Node root, int val){
        if(root.data > val){
            root.left= deleteEle(root.left,val);
        }else if(root.data<val){
            root.right= deleteEle(root.right,val);
        }else{
            // case1
            if(root.left == null && root.right==null){
                return null;
            }
            if(root.left==null){
                return root.right;
            }else if(root.right==null){
                return root.left;
            }
            // case 3
            Node IS = inOrderSuccessor(root.right);
            root.data=IS.data;
            root.right=deleteEle(root.right,IS.data);
        }
        return root;
    }
    public static Node inOrderSuccessor(Node root){
        while(root.left !=null){
            root = root.left;
        }
        return root;
    }
    public static void main(String[] args) {
        int values[] = {5, 1, 3, 4, 2, 7};
        Node root = null;
        for (int i = 0; i < values.length; i++) {
            root = insert(root, values[i]);
        }
            System.out.println("---------------");
        inOrder(root);
            System.out.println("");
            System.out.println("---------------");
        if(search(root,27)){
            System.out.println("Element Found");
        }else{
        System.out.println("Element Not Found");
        }
        deleteEle(root,1);
        inOrder(root);
    }
}
 

File: Tree/BST/BST.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Node
{
    int data;
    struct Node *left;
    struct Node *right;
} Node;
Node *createNode(int data)
{
    Node *newNode = (Node *)malloc(sizeof(Node));
    newNode->data = data;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}
Node *insert(Node *root, int val)
{
    if (root == NULL)
    {
        root = createNode(val);
        return root;
    }
    if (val < root->data)
    {
        root->left = insert(root->left, val);
    }
    else
    {
        root->right = insert(root->right, val);
    }
    return root;
}

void inOrder(Node *root)
{
    if (root == NULL)
        return;
    inOrder(root->left);
    printf("%d ", root->data);
    inOrder(root->right);
}
int search(Node *root, int key)
{
    if (root == NULL)
        return 0;
    if (root->data > key)
    {
        return search(root->left, key);
    }
    else if (root->data == key)
    {
        return 1;
    }
    else
    {
        return search(root->right, key);
    }
}
Node *inOrderSuccessor(Node *root)
{
    while (root->left != NULL)
        root = root->left;

    return root;
}
Node *deleteNode(Node *root, int val)
{
    if (root == NULL)
        return root;

    if (val < root->data)
        root->left = deleteNode(root->left, val);

    else if (val > root->data)
        root->right = deleteNode(root->right, val);

    else
    {
        // Case 1: No child
        if (root->left == NULL && root->right == NULL)
        {
            free(root);
            root = NULL;
        }
        // Case 2: One child (right)
        else if (root->left == NULL)
        {
            Node *temp = root;
            root = root->right;
            free(temp);
        }
        // Case 2: One child (left)
        else if (root->right == NULL)
        {
            Node *temp = root;
            root = root->left;
            free(temp);
        }
        // Case 3: Two children
        else
        {
            Node *temp = inOrderSuccessor(root->right);
            root->data = temp->data;
            root->right = deleteNode(root->right, temp->data);
        }
    }
    return root;
}

int main()
{
    int values[] = {5, 1, 3, 4, 2, 7};
    Node *root = NULL;
    for (int i = 0; i < sizeof(values) / sizeof(values[0]); i++)
    {
        root = insert(root, values[i]);
    }
    printf("---------------\n");
    inOrder(root);
    printf("\n");
    deleteNode(root, 3);
    printf("In-Order traversal after deletion:\n");
    inOrder(root);
    printf("\n");
    printf("---------------\n");
    if (search(root, 27))
    {
        printf("Element Found\n");
    }
    else
    {
        printf("Element Not Found\n");
    }

    return 0;
}